### PR TITLE
HOSTEDCP-788: Configurable SRE MetricsSet

### PR DESCRIPTION
**Pull request by csrwng**
**What this PR does / why we need it**:
Introduces the use of a ConfigMap to configure the metrics relabelings to use in control plane ServiceMonitors. The ConfigMap should live in the hypershift operator namespace, be named 'sre-metric-set' and have a single key named 'config'. The contents of the config should be a YAML file with keys for each of the control plane components that produces a service monitor. An example is included in support/metrics/testdata.

Changes to the configuration ConfigMap result in the ConfigMap getting copied to all hosted control planes and the Control Plane Operator pod in those namespaces to be restarted. For the CPO pod, the configuration is only read once during startup and used by the service monitor reconciliation functions.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #[HOSTEDCP-788](https://issues.redhat.com//browse/HOSTEDCP-788)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

**Review by sjenning — COMMENTED** [review]
looks good, just one question

File: hypershift-operator/main.go [hunk 1]
```diff
@@ -228,6 +228,30 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
 	}
 	log.Info("Using metrics set", "set", metricsSet.String())
 
+	sreConfigHash := ""
+	if metricsSet == metrics.MetricsSetSRE {
+		kubeClient, err := kubernetes.NewForConfig(mgr.GetConfig())
+		if err != nil {
+			log.Error(err, "unable to create kubernetes client")
+			os.Exit(1)
```
[the remark this answers is not in the document]
Comment (csrwng) [reply hypershift-operator/main.go:236]: true, will fix

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 2]
```diff
@@ -4716,6 +4753,38 @@ func (r *HostedClusterReconciler) discoverKubevirtClusterClient(ctx context.Cont
...
+		if apierrors.IsNotFound(err) {
+			log.Info("SRE configuration does not exist")
+			return nil
+		}
+		return fmt.Errorf("failed to get SRE configuration configmap: %w", err)
+	}
+	currentMetricsSetConfigHash := metrics.SREMetricsSetConfigHash(cm)
+	if currentMetricsSetConfigHash != r.SREConfigHash {
+		// Only load a new config if configuration content has changed
+		if err := metrics.LoadSREMetricsSetConfigurationFromConfigMap(cm); err != nil {
```
[the remark this answers is not in the document]
Comment (csrwng) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4768]: because the singleton variable in support/metrics/sets.go needs to be updated with the latest SRE config. From the HO we create the service monitor for the cpo, so it needs to be available for this process as well in case you want to update that config.

File: control-plane-operator/main.go [hunk 3]
```diff
@@ -361,6 +364,32 @@ func NewStartCommand() *cobra.Command {
 			os.Exit(1)
 		}
 		setupLog.Info("Using metrics set", "set", metricsSet.String())
+
+		sreConfigHash := ""
+		if metricsSet == metrics.MetricsSetSRE {
+			readerClient := mgr.GetAPIReader()
```
Comment (enxebre) [inline control-plane-operator/main.go:370]: why is this needed in main? Isn't it redundant with the main reconciliation https://github.com/openshift/hypershift/blob/main/test/e2e/util/fixture.go#L240-L247C5?

File: hypershift-operator/main.go [hunk 4]
```diff
@@ -228,6 +227,25 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
 	}
 	log.Info("Using metrics set", "set", metricsSet.String())
 
+	sreConfigHash := ""
+	if metricsSet == metrics.MetricsSetSRE {
```
Comment (enxebre) [inline hypershift-operator/main.go:231]: why is this needed in main? Isn't it redundant with the main reconciliation?

**Comment by enxebre** [comment]
verify failing
```
control-plane-operator/main.go:36:2: package "sigs.k8s.io/controller-runtime/pkg/client" is being imported more than once (ST1019)
	control-plane-operator/main.go:37:2: other import of "sigs.k8s.io/controller-runtime/pkg/client"
```

How can we prove somehow this work? May be include some of the metrics here https://github.com/openshift/hypershift/blob/main/test/e2e/util/fixture.go#L239-L247

Then follow up to handle the CM creation and check rules apply to final metric?

lgtm

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 5]
```diff
@@ -893,6 +894,11 @@ func IsStorageAndCSIManaged(hostedControlPlane *hyperv1.HostedControlPlane) bool
 }
 
 func (r *HostedControlPlaneReconciler) reconcile(ctx context.Context, hostedControlPlane *hyperv1.HostedControlPlane, createOrUpdate upsert.CreateOrUpdateFN, releaseImageProvider *imageprovider.ReleaseImageProvider, infraStatus InfrastructureStatus) error {
+
+	if err := r.reconcileSREMetricsConfig(ctx, createOrUpdate, hostedControlPlane.Namespace); err != nil {
```
Comment (bryan-cox) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:898]: Do we want to block all of the other reconciliation tasks done for the metrics?

**Review by bryan-cox — COMMENTED** [review]
Should we add function definition comments for all the newly added functions?

**Comment by csrwng** [comment]
> How can we prove somehow this work? May be include some of the metrics here https://github.com/openshift/hypershift/blob/main/test/e2e/util/fixture.go#L239-L247

We would have to change the default metric set for the e2e mgmt cluster to SRE (which may not be a bad thing to do), but it seems like something we should tackle in a separate, follow up PR.

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 5 above)
Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:898]: Yes, I would say that if we fail to load the configuration then whatever we reconcile will be wrong.

**Comment by csrwng** [comment]
> Should we add function definition comments for all the newly added functions?

Done
